File: Server/src/main/java/com/github/dr/rwserver/net/ConnectionAgreement.java
package com.github.dr.rwserver.net;

import com.github.dr.rwserver.io.Packet;
import com.github.dr.rwserver.net.udp.ReliableSocket;
import io.netty.channel.Channel;

import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Objects;

import static com.github.dr.rwserver.util.RandomUtil.generateStr;

/**
 * @author Dr
 * @date 2021年2月2日星期二 06:31:11
 */
public class ConnectionAgreement {
    private final ProtocolType<Packet> protocolType;
    private final Object object;
    private final DataOutputStream socketStream;
    public final String useAgreement;
    public final String ip;
    public final int localPort;
    public final String id;

    /**
     * TCP Send
     * @param channel Netty-Channel
     */
    public ConnectionAgreement(Channel channel) {
        protocolType = channel::writeAndFlush;
        object = channel;
        useAgreement = "TCP";
        ip = convertIp(channel.remoteAddress().toString());
        localPort = ((InetSocketAddress) channel.localAddress()).getPort();
        socketStream = null;
        id = null;
    }

    /**
     * UDP Send
     * @param socket Socket
     * @throws IOException Error
     */
    public ConnectionAgreement(Socket socket) throws IOException {
        socketStream = new DataOutputStream(socket.getOutputStream());
        protocolType = (msg) -> {
            socketStream.writeInt(msg.bytes.length);
            socketStream.writeInt(msg.type);
            socketStream.write(msg.bytes);
            socketStream.flush();
        };
        object = socket;
        useAgreement = "UDP";
        ip = convertIp(socket.getRemoteSocketAddress().toString());
        localPort = socket.getLocalPort();
        id = generateStr(5);
    }

    public void add(final GroupNet groupNet) {
        if (object instanceof Channel) {
            groupNet.add((Channel) object);
        } else if (object instanceof ReliableSocket) {
            groupNet.add(this);
        }
    }

    public void send(Packet packet) throws IOException {
        protocolType.send(packet);
    }

    public void close(final GroupNet groupNet) throws IOException {
        if (groupNet != null) {
            if (object instanceof Channel) {
                ((Channel) object).close();
                groupNet.remove((Channel) object);
                ((Channel) object).close();
            } else if (object instanceof ReliableSocket) {
                socketStream.close();
                groupNet.remove(this);
                ((ReliableSocket) object).close();
            }
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        return this.id.equals(((ConnectionAgreement) o).id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    private String convertIp(final String ipString) {
        return ipString.substring(1, ipString.indexOf(':'));
    }

    private interface ProtocolType<T>{
        /**
         * 接管Send逻辑
         * @param t Msg
         * @throws IOException Error
         */
        void send(T t) throws IOException;
    }
}

File: Server/src/main/java/com/github/dr/rwserver/net/core/ConnectionAgreement.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package com.github.dr.rwserver.net.core

import com.github.dr.rwserver.io.packet.Packet
import com.github.dr.rwserver.net.GroupNet
import com.github.dr.rwserver.net.rudp.PackagingSocket
import com.github.dr.rwserver.struct.Seq
import com.github.dr.rwserver.util.log.Log
import io.netty.channel.ChannelHandlerContext
import java.io.DataOutputStream
import java.io.IOException
import java.net.InetSocketAddress
import java.net.SocketException
import java.util.*

/**
 * @author Dr
 * @date 2021年2月2日星期二 06:31:11
 */
class ConnectionAgreement {
    private val protocolType: ((packet: Packet) -> Unit)
    private val objectOutStream: Any
    private val udpDataOutputStream: DataOutputStream?

    val isClosed: ()->Boolean
    val useAgreement: String
    @JvmField
    val ip: String
    internal val localPort: Int
    val id: String = UUID.randomUUID().toString()

    /**
     * TCP Send
     * @param channelHandlerContext Netty-ChannelHandlerContext
     */
    internal constructor(channelHandlerContext: ChannelHandlerContext) {
        protocolType = { packet: Packet ->
            channelHandlerContext.writeAndFlush(packet)
        }
        objectOutStream = channelHandlerContext
        udpDataOutputStream = null
        useAgreement = "TCP"
        isClosed = { false }

        val channel = channelHandlerContext.channel()
        ip = convertIp(channel.remoteAddress().toString())
        localPort = (channel.localAddress() as InetSocketAddress).port
    }

    /**
     * UDP Send
     * @param socket Socket
     * @throws IOException Error
     */
    internal constructor(socket: PackagingSocket) {
        val socketStream = DataOutputStream(socket.outputStream)
        protocolType = { packet: Packet ->
            socketStream.writeInt(packet.bytes.size)
            socketStream.writeInt(packet.type)
            socketStream.write(packet.bytes)
            socketStream.flush()
        }
        objectOutStream = socket
        udpDataOutputStream = socketStream
        useAgreement = "UDP"
        isClosed = { socket.isClosed }

        ip = convertIp(socket.remoteSocketAddressString)
        localPort = socket.localPort
    }

    constructor() {
        protocolType = {}
        objectOutStream = ""
        udpDataOutputStream = null
        useAgreement = "Test"
        isClosed = { false }

        ip = ""
        localPort = 0
    }

    fun add(groupNet: GroupNet) {
        if (objectOutStream is ChannelHandlerContext) {
            groupNet.add(objectOutStream.channel())
        } else if (objectOutStream is PackagingSocket) {
            groupNet.add(this)
        }
    }

    /**
     * 接管Send逻辑
     * 整合不同协议的发送逻辑
     * @param packet MsgPacket
     * @throws IOException Error
     */
    @Throws(IOException::class)
    fun send(packet: Packet) {
        protocolType(packet)
    }

    @Throws(IOException::class)
    fun close(groupNet: GroupNet?) {
        IPData.remove(ip)
        if (groupNet != null) {
            if (objectOutStream is ChannelHandlerContext) {
                groupNet.remove(objectOutStream.channel())
            } else if (objectOutStream is PackagingSocket) {
                groupNet.remove(this)
            }
        }
        if (objectOutStream is ChannelHandlerContext) {
            objectOutStream.channel().close()
            objectOutStream.close()
        } else if (objectOutStream is PackagingSocket) {
            try {
                udpDataOutputStream!!.close()
                objectOutStream.close()
            } catch (e : SocketException) {
                Log.debug("[RUDP Close] Passive")
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        return if (other == null || javaClass != other.javaClass) {
            false
        } else id == (other as ConnectionAgreement).id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    private fun convertIp(ipString: String): String {
        return ipString.substring(1, ipString.indexOf(':'))
    }

    companion object {
        val IPData = Seq<String>()
    }
}